include/gazebo_magnetometer_plugin.h

```diff
@@ -0,0 +1,110 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2019 PX4 Development Team. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+/**
+ * @brief Magnetometer Plugin
+ *
+ * This plugin simulates magnetometer data
+ *
+ * @author Elia Tarasov <elias.tarasov@gmail.com>
+ */
+
+#ifndef _GAZEBO_MAGNETOMETER_PLUGIN_HH_
+#define _GAZEBO_MAGNETOMETER_PLUGIN_HH_
+
+#include <random>
+#include <string>
+
+#include <Eigen/Core>
+
+#include <Magnetometer.pb.h>
+#include <Groundtruth.pb.h>
+
+#include <gazebo/common/common.hh>
+#include <gazebo/common/Plugin.hh>
+#include <gazebo/gazebo.hh>
+#include <gazebo/physics/physics.hh>
+#include <gazebo/transport/transport.hh>
+#include <gazebo/msgs/msgs.hh>
+
+#include <ignition/math.hh>
+
+#include <common.h>
+
+#include <geo_mag_declination.h>
+
+namespace gazebo {
+
+static constexpr auto kDefaultMagnetometerTopic = "mag";
+static constexpr auto kDefaultPubRate = 20.0; // [Hz]
+static constexpr auto kDefaultNoiseDensity = 0.01; // [gauss / sqrt(hz)]
```

Comment:
Is this noise model correct for mag field data?

Reply:
I'm not sure. This is the default [model](https://github.com/PX4/sitl_gazebo/blob/master/src/gazebo_mavlink_interface.cpp#L613). Maybe we need approach similar to the `gazebo_imu_plugin` or default ROS sensor model.

Reply:
Any suggestions on that?

Reply:
@priseborough can you take a quick look at this?